### PR TITLE
[DO NOT MERGE] Jane's autograd lab pts 1+2

### PR DESCRIPTION
**Pull request by janeyx99**
Instructions from https://github.com/pytorch/pytorch/wiki/Autograd-Onboarding-Lab

Much code copied from LinearFunction example in https://pytorch.org/docs/stable/notes/extending.html

### CODE REVIEW

File: test/attn.py [hunk 1]
```diff
@@ -0,0 +1,72 @@
...
+
+    # Note that both forward and backward are @staticmethods
+    @staticmethod
+    # bias is an optional argument
+    def forward(ctx, q, k, v):
+        assert len(q.size()) == 2, f"input q must be 2D but instead has {len(q.size())} dims"
+        assert len(k.size()) == 2, f"input k must be 2D but instead has {len(k.size())} dims"
+        assert len(v.size()) == 2, f"input v must be 2D but instead has {len(v.size())} dims"
+        assert q.size(0) == k.size(0) and k.size(0) == v.size(0), "all inputs must have the same first dimension"
+        assert q.size(1) == k.size(1), "q and k must share the same size for their second dimension"
```
Comment (janeyx99) [inline test/attn.py:22]: Should probably add `ctx.set_materialize_grads(False)` here

File: test/attn.py [hunk 2]
```diff
@@ -0,0 +1,72 @@
+from torch import autograd
+import torch
+
+# Inherit from Function
+class AttentionFunction(autograd.Function):
+    @staticmethod
+    def attn(ctx, q, k, v):
```
Comment (soulitzer) [inline test/attn.py:7]: interesting that you would split this out, typically we'd just have this logic inside the forward

File: test/attn.py [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+        x = torch.matmul(q, k.transpose(0, 1))
+        a = torch.tanh(x)
+        ctx.save_for_backward(q, k, v, x, a)
```
Comment (soulitzer) [inline test/attn.py:10]: since you saved `a` here you can replace usages of `tanh(x)` with that

Comment (soulitzer) [reply test/attn.py:10]: if you want your function to support double backward, you should only be saving the inputs and outputs your function. Because gradient tracking is disabled when forward is run, intermediate values computed during forward will not have gradients tracked. So, if we use them during the backward the contributions of the intermediary values to the outputs won't be properly added and the gradients would be wrong.

More context: https://pytorch.org/tutorials/intermediate/custom_function_double_backward_tutorial.html

Comment (janeyx99) [reply test/attn.py:10]: :o right!

File: test/attn.py [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+            grad_oq = ((grad_o @ v.transpose(0, 1)) * (1 - torch.tanh(x) ** 2)) @ k
+            grad_aq = (grad_a * (1 - torch.tanh(x) ** 2)) @ k
+            total_q = grad_oq + grad_aq
+        if ctx.needs_input_grad[1]:
+            # I used to have these be transposed at the end but that didn't pass the tests
+            grad_ok = ((grad_o @ v.transpose(0, 1)) * (1 - torch.tanh(x) ** 2)).transpose(0, 1) @ q
+            grad_ak = (grad_a * (1 - torch.tanh(x) ** 2)).transpose(0, 1) @ q
+            total_k = grad_ok + grad_ak
+        if ctx.needs_input_grad[2]:
+            total_v = a.transpose(0, 1) @ grad_o
```
Comment (soulitzer) [inline test/attn.py:50]: if you want to optimize this further things you can do:
- do in-place ops, but don't mutate inputs or saved tensors
- some common subexpressions can probably be factored out, i.e. torch.tanh(x) ** 2

File: test/attn.py [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+'''
+
+# gradcheck takes a tuple of tensors as input, check if your gradient
+# evaluated with these tensors are close enough to numerical
+# approximations and returns True if they all verify this condition.
+q = torch.rand(2, 3, requires_grad=True, dtype=torch.float64)
+k = torch.rand(2, 3, requires_grad=True, dtype=torch.float64)
+v = torch.rand(2, 4, requires_grad=True, dtype=torch.float64)
+input = (q, k, v)
+test = autograd.gradcheck(AttentionFunction.apply, input, eps=1e-6, atol=1e-4)
```
Comment (soulitzer) [inline test/attn.py:71]: do the default eps and atol work?

File: test/attn.py [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+        # optional inputs.
+        q, k, v, x, a = ctx.saved_tensors
+        total_q = total_k = total_v = None
+
+        # These needs_input_grad checks are optional and there only to
+        # improve efficiency. If you want to make your code simpler, you can
+        # skip them. Returning gradients for inputs that don't require it is
+        # not an error.
+        if ctx.needs_input_grad[0]:
+            grad_oq = ((grad_o @ v.transpose(0, 1)) * (1 - torch.tanh(x) ** 2)) @ k
```
Comment (soulitzer) [inline test/attn.py:41]: x.T is shorthand for transpose(0, 1) when x is 2D

Comment (janeyx99) [reply test/attn.py:41]: :O

File: test/attn.py (hunk 4 above)
Comment (janeyx99) [reply test/attn.py:50]: i was def not thinking of optimization then, but good points

File: test/attn.py (hunk 5 above)
Comment (janeyx99) [reply test/attn.py]: yes

File: test/attn.py [hunk 7]
```diff
@@ -0,0 +1,70 @@
...
+
+        # These needs_input_grad checks are optional and there only to
+        # improve efficiency. If you want to make your code simpler, you can
+        # skip them. Returning gradients for inputs that don't require it is
+        # not an error.
+
+        i = grad_o @ v.T if grad_o is not None else torch.zeros(1)
+        j = 1 - a ** 2
+        if ctx.needs_input_grad[0]:
+            total_q = (i * j) @ k + ((grad_a * j) @ k if grad_a is not None else torch.zeros(1))
```
Comment (soulitzer) [inline test/attn.py:44]: A few more small comments (no action needed!):
- FYI another way we represent zeros with a special object called ZeroTensor which reduces unnecessary computation when doing certain operations like multiplying or adding See https://github.com/pytorch/pytorch/issues/69687 for more context (currently it is mostly used for forward AD)
- For this case specifically I would just conditionally add the second term `(grad_a * j) @ k` if it is defined to avoid materializing that extra zero
- naming nit: It might be better to use upper case letters like `A`, `V`, `Q`, to refer to 2-D tensors (matrices), vs i, j which tends to refer to indices

Comment (janeyx99) [reply test/attn.py:44]: acknowledged! zerotensors seem cool

**Comment by janeyx99** [comment]
Closing this to mark this portion as complete :D
